component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+func (rw *ReadWriteFilenameCache) DeleteDirectory(options internal.DeleteDirOptions) error {
+	log.Trace("Stream::DeleteDirectory : name=%s", options.Name)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Name) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().DeleteDir(options)
+	if err != nil {
+		log.Err("Stream::DeleteDirectory : error deleting directory %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) RenameDirectory(options internal.RenameDirOptions) error {
+	log.Trace("Stream::RenameDirectory : name=%s", options.Src)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Src) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().RenameDir(options)
+	if err != nil {
+		log.Err("Stream::RenameDirectory : error renaming directory %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (rw *ReadWriteFilenameCache) Stop() error {
+	log.Trace("Stopping component : %s", rw.Name())
+	if !rw.StreamOnly {
+		rw.Lock()
+		defer rw.Unlock()
+		for fileName, buffer := range rw.fileCache {
+			delete(rw.fileCache, fileName)
+			buffer.Lock()
+			defer buffer.Unlock()
+			buffer.Purge()
+			atomic.AddInt32(&rw.CachedObjects, -1)
+		}
+	}
+	return nil
+}
+
+// GetAttr : Try to serve the request from the attribute cache, otherwise cache attributes of the path returned by next component
+func (rw *ReadWriteFilenameCache) GetAttr(options internal.GetAttrOptions) (*internal.ObjAttr, error) {
+	// log.Trace("AttrCache::GetAttr : %s", options.Name)
+	attrs, err := rw.NextComponent().GetAttr(options)
+	if err != nil {
+		log.Err("Stream::GetAttr : error getting attributes %s [%s]", options.Name, err.Error())
+		return nil, err
+	}
+	rw.RLock()
+	defer rw.RUnlock()
+	buffer, found := rw.fileCache[options.Name]
+	if !found {
+		return attrs, err
+	}
+	attrs.Mtime = buffer.Mtime
+	attrs.Size = buffer.Size
+	return attrs, nil
+}
+
+func (rw *ReadWriteFilenameCache) purge(fileName string, close bool) {
+	// check if this file is cached
+	rw.Lock()
+	defer rw.Unlock()
+	buffer, found := rw.fileCache[fileName]
+	if found {
+		// if it is a close operation then decrement the handle count on the buffer
+		if close {
+			atomic.AddInt64(&buffer.HandleCount, -1)
+		}
+		// rw.RUnlock()
+		// if the handle count is 0 (no open handles) purge the buffer
+		if atomic.LoadInt64(&buffer.HandleCount) <= 0 || !close {
+			delete(rw.fileCache, fileName)
+			buffer.Lock()
+			defer buffer.Unlock()
+			buffer.Purge()
+			buffer.StreamOnly = true
+			atomic.AddInt32(&rw.CachedObjects, -1)
+		}
+	}
+}
+
+//TODO: Store block offset list
+func (rw *ReadWriteFilenameCache) createFileCache(handle *handlemap.Handle) error {
+	// check if file is cached
+	rw.Lock()
```

Comment:
This function is doing a lot of operations, do we really need this lock to be up for entire lifespan of this method.